### DISCUSSION TITLE
useFormEffects 注册的 effect 为啥不能销毁？

### DISCUSSION THREAD

**Discussion by javahuang** [alibaba/formily#1489] [Q&A 2.x] [answered]
代码见：[codesandbox](https://codesandbox.io/s/solitary-cache-t4xo7?file=/App.tsx)

简单来说，在 Field component 内部使用 FormEffects，每次 Field remount 的时候都会重新注册一个 effect，并且之前注册的 effect 不会销毁，导致回调执行多次。

```ts
const Input: React.FC<InputProps> = (props) => {
  useFormEffects(() => {
    onFieldValueChange("aa", (f) => {
      console.log(f.value);
    });
  });

  return <input {...props} value={props.value || ""} />;
};
```

我看 useFormEffects 内部在 field unmount 的时候会删除掉该 effect 的，不知道为何没有生效。

https://github.com/alibaba/formily/blob/91b4b617a115b80bd6debb8d43c36d10a749c153/packages/react/src/hooks/useFormEffects.ts#L8-L18

之前讨论过，@janryWang  的回复如下

>useFormEffects是往全局注册effects的，它是感知不到当前组件的路径是哪个，所以每次新实例出现，它都会重新监听

所以想了解下细节，useFormEffects 销毁 effect 这块没啥没有生效？

**Comment by janryWang** [comment, marked as answer]
具体怎么复现的

**Reply by javahuang** [reply]
在示例里面每次点击 hide 之后再 show，effect 就会多注册一次，当 input 的 value 变化时， onFieldValueChange 就会执行多次。
但是刚试了 codesandbox 好像又没问题，我本地用 vite 起了一个项目，相同的代码，却一直是多次。

**Reply by janryWang** [reply]
那你先排查一下吧

**Reply by javahuang** [reply]
我刚本地用 create-react-app 起了一个项目，和 vite 启动结果是一样的。
在 Field 内部通过 useFormEffects 注册的 effect 不会销毁。

**Reply by janryWang** [reply]
你试试在组件里写个useEffect，看看会不会触发unmount

**Reply by javahuang** [reply]
会触发

**Reply by janryWang** [reply]
是不是用了StrictMode

**Reply by javahuang** [reply]
非常感谢，真的是和这个有关系，请问为啥开启 StrictMode 会导致这个问题了？

**Reply by janryWang** [reply]
StrictMode会触发多次useMemo执行，这个很恶心

**Reply by janryWang** [reply]
其实这个问题是可以修掉的，就是比较绕，你可以提个issue，后面我处理一下
